### PR TITLE
[DO NOT MERGE FOR NOW] Dont use discovery topic

### PR DESCRIPTION
**Pull request by cammellos**
Moves to public key topic instead of discovery.

This will reduce the amount of messages that needs to be fetched from the mailserver.

Having a single topic eventually will not scale, from the client perspective, as it scales with the number of clients in the network (total number of contact messages exchanged), moving to a public key topic scales based on how much activity there is for a given user.

The tradeoff is that we need to also listen to the topic we publish, in order to avoid giving away who we are communicating with (light nodes do not forward messages that are not in bloom filter).

This is a breaking change, users with older version of the app will not be able to communicate with this version.

This is only a performance improvement so no functionality should be fixed nor broken.

status: ready

### CODE REVIEW

File: src/status_im/transport/filters.cljs
Comment (rasom) [inline src/status_im/transport/filters.cljs:31]: Why can't this be done in callback?

Comment (cammellos) [reply src/status_im/transport/filters.cljs:31]: I was wondering the same, but decided not to change it to keep code changes to a minimum. The callback version would take 4 arguments https://web3js.readthedocs.io/en/1.0/web3-shh.html#shh-subscribeoptions (subscribe has the same signature as  newmessagefilter), with the last being the returned value, I can try and see if it works.

**Review by pedropombeiro — COMMENTED** [review]
Is there any documentation that needs to be updated to reflect these changes, e.g. https://docs.google.com/document/d/1Qh2h07T_qepzEJ7IytmxwIdQAOsGHrvhXwZxuZtbwgc/edit?ts=5ac2d6fe ?

**Comment by cammellos** [comment]
@PombeirP yes, that would need to be updated, good call, thanks

File: src/status_im/data_store/realm/schemas/account/core.cljs
Comment (pedropombeiro) [inline src/status_im/data_store/realm/schemas/account/core.cljs]: Complete newbie comment here, but is this `schemaVersion` supposed to stay at 1?

Comment (cammellos) [reply src/status_im/data_store/realm/schemas/account/core.cljs]: no, correct, needs to be bumped to version 2, I have the change locally, but haven't pushed

File: src/status_im/transport/handlers.cljs
Comment (janherich) [inline src/status_im/transport/handlers.cljs]: @cammellos Are you sure we don't want this information (`:one-to-one` flag set to true) persisted as well ?
You establish the filter with it, but don't include it in the `chat-transport-info` map which is persisted, so the next time app starts, I'm afraid the filter will be established without that flag.

**Review by pedropombeiro — APPROVED** [review]
Can't provide feedback on the Clojure code changes, but I think the spirit of the change makes sense.

File: src/status_im/transport/handlers.cljs
Comment (cammellos) [reply src/status_im/transport/handlers.cljs]: @janherich I will double check that to make sure, but I think it is persisted, I have added it in `create-chat` above, defaults to false but allow overriding (we set it to `true` for public chats)

Comment (cammellos) [reply src/status_im/transport/handlers.cljs]: Here, https://github.com/status-im/status-react/blob/ac48f9edac136cca91b489783c8c691b382b34d5/src/status_im/transport/db.cljs#L32

**Comment by janherich** [comment]
@cammellos What's the rationale of having `:filters` collection, instead of just one `:filter` for each transport (identified by `:chat-id` entry) ?

File: src/status_im/transport/handlers.cljs
Comment (janherich) [reply src/status_im/transport/handlers.cljs]: Ok, I see, it was already set correctly by previously and this just adds the filter.

**Comment by cammellos** [comment]
@janherich sure, this only applies for 1-to-1 btw:

When I want to send a contact request to you I am now using Hash(Janpk).
As discussed we also need to advertise the bloom filter, to avoid giving up our identity, so we add a noop filter for Hash(Janpk).
We also add a filter for the symmetric key we sent.

Once we move to publishing to pk we have a choice to have a single filter Hash(Janpk) or keep also the symmetric-key (I think there are some arguments to keep both filters).

**Comment by cammellos** [comment]
The assumption is also that we would give up our identity if we just listen to the filter once we publish, so we want to keep the noop filter up as long as there is an open "session".

**Comment by janherich** [comment]
@cammellos Ok, so the reason is to have both generated uuid topic/sym key combination (I thought we abandoned that for 1-1 chats for now) and user-pk/asymkey combination as filters.

**Review by janherich — APPROVED** [review]

File: src/status_im/transport/handlers.cljs
Comment (cammellos) [reply src/status_im/transport/handlers.cljs]: @janherich correct, this applies to the current develop, if we merge the changes to publish to pk this will have to be revisited.

File: src/status_im/transport/filters.cljs
Comment (cammellos) [reply src/status_im/transport/filters.cljs:31]: no dice, callback does not fire on success, only on error

Comment (cammellos) [reply src/status_im/transport/filters.cljs:31]: @rasom could you have another look at the PR? I had to rebase against the linting changes and I don't want to be the first one to introduce badly formatted code :)

Comment (janherich) [reply src/status_im/transport/filters.cljs:31]: @cammellos One area where I found your formatting not following the clojure styleguide are the comments, you use `;` comment prefix for standalone comment lines, which is wrong (`;` is only to be used for after the line comments, which we agreed on not to use), `;;` should be used instead.
Probably not related to this PR, more like an general remark.

Comment (cammellos) [reply src/status_im/transport/filters.cljs:31]: @janherich guilty as charged, I didn't know the difference, thanks!

**Comment by jeluard** [comment]
@cammellos Can you please rebase? Thanks!

**Comment by cammellos** [comment]
@jeluard rebased, thanks

**Comment by jeluard** [comment]
@cammellos Actually you can squash your commits too! :)

**Comment by cammellos** [comment]
@jeluard done :)

**Comment by lukaszfryc** [comment]
There are no regressions based on my testing. Now, let me compare the bandwith consumption with nightly build.

**Comment by lukaszfryc** [comment]
@cammellos for a test where I send 100 messages and receive 100 messages in 1-1 chat in foreground, the bandwidth consumption look pretty similar between this build and nightly (7233a5). 

I've run each test 2 times per build. Results:

|Test No|Build|Data transferred|Power usage|
|---|---|---|---|
|1|PR|4.19 MB total (0.86 MB received, 3.34 MB transmitted)|0.64%|
|2|PR|4.47 MB total (0.95 MB received, 3.51 MB transmitted)|0.65%|
|3|Nightly|4.25 MB total (0.89 MB received, 3.36 MB transmitted)|0.72%|
|4|Nightly|3.95 MB total (0.91 MB received, 3.04 MB transmitted)|0.62%|

Let me know if you think we should do other kind of tests. If not, please merge.

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/nrr76T for branch PR-4135

<blockquote><div><strong><a href="https://i.diawi.com/nrr76T">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/sHMZFM for branch PR-4135

<blockquote><div><strong><a href="https://i.diawi.com/sHMZFM">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/bhXZZv for branch PR-4135

<blockquote><div><strong><a href="https://i.diawi.com/bhXZZv">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/9FJEQS for branch PR-4135

<blockquote><div><strong><a href="https://i.diawi.com/9FJEQS">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by oskarth** [comment]
@cammellos @lukaszfryc What happened with this PR? Has it been established what the perf increase is?

**Comment by cammellos** [comment]
@oskarth At the moment seems like the performance increase is not that relevant as most of the people use public chats to communicate. 

Because this method trade-offs some obscurity, I would wait until other performance improving PRs are merged (mailserver limit & from), and re-evaluate. But otherwise we are ready to pull the trigger, has been tested & reviewed.

**Pull request closed without merging**
